<h2 style="color:orange;">While loop</h2>

<p>The while loop is like a repeated if statement. The code is executed over and over again, as long as the condition is True. Have another look at its recipe.

while condition :<br>
    expression<br>
Can you tell how many printouts the following while loop will do?<br>

x = 1<br>
while x < 4 :<br>
    print(x)<br>
    x = x + 1</p><br>

   

In [1]:
x = 1
while x < 4 :
    print(x)
    x = x + 1

1
2
3


 Correct! After 3 runs, x will be equal to 4, causing x < 4 to evaluate to False. This means that the while loop is executed 3 times, giving three printouts.

<li>Create the variable offset with an initial value of 8.</li>
<li>Code a while loop that keeps running as long as offset is not equal to 0. Inside the while loop:</li>
<li>Print out the sentence "correcting...".</li>
<li>Next, decrease the value of offset by 1. You can do this with offset = offset - 1.</li>
<li>Finally, still within your loop, print out offset so you can see how it changes.</li>

In [2]:
# Initialize offset
offset = 8


# Code the while loop
while offset != 0:
    print('correcting...')
    offset =offset - 1
    print(offset)

correcting...
7
correcting...
6
correcting...
5
correcting...
4
correcting...
3
correcting...
2
correcting...
1
correcting...
0


Inside the while loop, complete the if-else statement:<br>
If offset is greater than zero, you should decrease offset by 1.<br>
Else, you should increase offset by 1.<br>
If you've coded things correctly, hitting Submit Answer should work this time.<br>
If your code is still taking too long to run (or your session is expiring), you probably made a mistake. Check your code and make sure that the statement offset != 0 will eventually evaluate to FALSE!<br>

In [3]:
# Initialize offset
offset = -6

# Code the while loop
while offset != 0 :
    print("correcting...")
    if offset > 0 :
      offset=offset-1
    else : 
      offset=offset+1
    print(offset)

correcting...
-5
correcting...
-4
correcting...
-3
correcting...
-2
correcting...
-1
correcting...
0


<span style="color:orange;font-weight:bold;">Loop over the list</span>
<p><br>
fam = [1.73, 1.68, 1.71, 1.89]<br>
for height in fam : <br>
    print(height)<br>
As usual, you simply have to indent the code with 4 spaces to tell Python which code should be executed in the for loop.<br>
The areas variable, containing the area of different rooms in your house, is already defined.</p>

<span>Write a for loop that iterates over all elements of the areas list and prints out every element separately.</span>

In [4]:
# areas list
areas = [11.25, 18.0, 20.0, 10.75, 9.50]

# Code the for loop
for area in areas:
    print(area)

11.25
18.0
20.0
10.75
9.5


<span style="color:orange;font-weight:bold;">Indexes and values</span>
<p>Using a for loop to iterate over a list only gives you access to every list element in each run, one after the other. If you also want to access the index information, so where the list element you're iterating over is located, you can use enumerate().</p>
<p>As an example, have a look at how the for loop from the video was converted:</p>
<p>fam = [1.73, 1.68, 1.71, 1.89]</p>
<p>for index, height in enumerate(fam) :</p>
<p>print("person " + str(index) + ": " + str(height))</p>

<li>Adapt the for loop in the sample code to use enumerate() and use two iterator variables.</li>
<li>Update the print() statement so that on each run, a line of the form "room x: y" should be printed, where x is the index of the list element and y is the actual list element, i.e. the area. Make sure to print out this exact string, with the correct spacing.</li>

In [6]:
# areas list
areas = [11.25, 18.0, 20.0, 10.75, 9.50]

# Change for loop to use enumerate() and update print()
for index,a in enumerate(areas) :
    print("room "+ str(index)+ ": "+ str(a))

room 0: 11.25
room 1: 18.0
room 2: 20.0
room 3: 10.75
room 4: 9.5


<p>For non-programmer folks, room 0: 11.25 is strange. Wouldn't it be better if the count started at 1?</p>
<p>Adapt the print() function in the for loop so that the first printout becomes "room 1: 11.25", the second one "room 2: 18.0" and so on.</p>

In [7]:
# Code the for loop
for index, area in enumerate(areas) :
    print("room " + str(index + 1) + ": " + str(area))

room 1: 11.25
room 2: 18.0
room 3: 20.0
room 4: 10.75
room 5: 9.5


<span style="color:orange;font-weight:bold;">Loop over list of lists</span><br>
It's up to you to build a for loop from scratch this time!<br>
Write a for loop that goes through each sublist of house and prints out the x is y sqm, where x is the name of the room and y is the area of the room.

In [8]:
# house list of lists
house = [["hallway", 11.25], 
         ["kitchen", 18.0], 
         ["living room", 20.0], 
         ["bedroom", 10.75], 
         ["bathroom", 9.50]]
         
# Build a for loop from scratch
for room in house:
    print("the " + str(room[0]) + " is " + str(room[1]) + " sqm")

the hallway is 11.25 sqm
the kitchen is 18.0 sqm
the living room is 20.0 sqm
the bedroom is 10.75 sqm
the bathroom is 9.5 sqm


<span style="color:orange;font-weight:bold;">Loop over Dictionaries</span><br>
<p>n Python 3, you need the items() method to loop over a dictionary:<br>

world = { "afghanistan":30.55, <br>
          "albania":2.77,<br>
          "algeria":39.21 }<br>

for key, value in world.items() :<br>
    print(key + " -- " + str(value))<br>
Remember the europe dictionary that contained the names of some European countries as key and their capitals as corresponding value? Go ahead and write a loop to iterate over it!</p>

<li>Write a for loop that goes through each key:value pair of europe. On each iteration, "the capital of x is y" should be printed out, where x is the key and y is the value of the pair.</li>

In [9]:
# Definition of dictionary
europe = {'spain':'madrid', 'france':'paris', 'germany':'berlin',
          'norway':'oslo', 'italy':'rome', 'poland':'warsaw', 'austria':'vienna' }
          
# Iterate over europe
for key, value in europe.items() :
    print("the capital of " + key + " is " + str(value))

the capital of spain is madrid
the capital of france is paris
the capital of germany is berlin
the capital of norway is oslo
the capital of italy is rome
the capital of poland is warsaw
the capital of austria is vienna


<span style="color:orange;font-weight:bold;">Loop over NumPy array</span><br>
<p>If you're dealing with a 1D NumPy array, looping over all elements can be as simple as:<br>

for x in my_array :<br>
    ...</p>

<p>If you're dealing with a 2D NumPy array, it's more complicated. A 2D array is built up of multiple 1D arrays. To explicitly iterate over all separate elements of a multi-dimensional array, you'll need this syntax:</p>
<p>for x in np.nditer(my_array) :<br>
    ...</p>



In [15]:
import numpy as np

# 1D array
array_1d = np.array([1, 2, 3, 4, 5])
print("loop over 1D array")

# Using np.nditer
for element in np.nditer(array_1d):
    print(element)

print()
# 2D array
array_2d = np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9]])
print("loop over 2D array using np.nditer")

# Using np.nditer
for element in np.nditer(array_2d):
    print(element)
print()
print("loop over 2D array using flatten")

# Using flatten() 
# Looping over flattened 2D array
for element in array_2d.flatten():
    print(element)


loop over 1D array
1
2
3
4
5

loop over 2D array using np.nditer
1
2
3
4
5
6
7
8
9

loop over 2D array using flatten
1
2
3
4
5
6
7
8
9


<span style="color:orange;font-weight:bold;">Loop over DataFrames</span><br>
<p>Iterating over a Pandas DataFrame is typically done with the iterrows() method. Used in a for loop, every observation is iterated over and on every iteration the row label and actual row contents are available:<br>

for lab, row in brics.iterrows() :<br>
    ...<br>
In this and the following exercises you will be working on the cars DataFrame. It contains information on the cars per capita and whether people drive right or left for seven countries in the world.</p>

<li>Write a for loop that iterates over the rows of cars and on each iteration perform two print() calls: one to print out the row label and one to print out all of the rows contents.</li>

In [22]:
# Import cars data
import pandas as pd
cars = pd.read_csv('C:/Users/RBTG/OneDrive/Desktop/Data science/data/cars.txt', index_col = 0)

# Iterate over rows of cars
for label,row in cars.iterrows():
    print(label)
    print()
    print(row)
    print("-------------------------------")

US

country         United States
drives_right             True
cars_per_cap              809
Name: US, dtype: object
-------------------------------
AUS

country         Australia
drives_right        False
cars_per_cap          731
Name: AUS, dtype: object
-------------------------------
JPN

country         Japan
drives_right    False
cars_per_cap      588
Name: JPN, dtype: object
-------------------------------
IN

country         India
drives_right    False
cars_per_cap       18
Name: IN, dtype: object
-------------------------------
RU

country         Russia
drives_right      True
cars_per_cap       200
Name: RU, dtype: object
-------------------------------
MOR

country         Morocco
drives_right       True
cars_per_cap         70
Name: MOR, dtype: object
-------------------------------
EG

country         Egypt
drives_right     True
cars_per_cap       45
Name: EG, dtype: object
-------------------------------


<p>The row data that's generated by iterrows() on every run is a Pandas Series. This format is not very convenient to print out. Luckily, you can easily select variables from the Pandas Series using square brackets:

for lab, row in brics.iterrows() :
    print(row['country'])</p>

<p>Using the iterators lab and row, adapt the code in the for loop such that<br> the first iteration prints out "US: 809",<br> the second iteration "AUS: 731", and so on.<br>
The output should be in the form "country: cars_per_cap". Make sure to print out this exact string (with the correct spacing).<br>
You can use str() to convert your integer data to a string so that you can print it in conjunction with the country label.</p>

In [23]:
# Adapt for loop
for lab, row in cars.iterrows() :
    print(str(lab) +": "+str(row["cars_per_cap"]))

US: 809
AUS: 731
JPN: 588
IN: 18
RU: 200
MOR: 70
EG: 45


<span style="color:orange;font-weight:bold;">Adding new Column</span><br>
<p>add the length of the country names of the brics DataFrame in a new column:<br>

for lab, row in brics.iterrows() :<br>
    brics.loc[lab, "name_length"] = len(row["country"])<br>
You can do similar things on the cars DataFrame.</p>


<li>Use a for loop to add a new column, named COUNTRY, that contains a uppercase version of the country names in the "country" column. You can use the string method upper() for this.</li>
<li>To see if your code worked, print out cars. Don't indent this code, so that it's not part of the for loop.</li>

In [24]:
# Code for loop that adds COUNTRY column
for lab,row in cars.iterrows():
    cars.loc[lab,"COUNTRY"]=row["country"].upper()


# Print cars
print(cars)

           country drives_right  cars_per_cap        COUNTRY
US   United States         True           809  UNITED STATES
AUS      Australia        False           731      AUSTRALIA
JPN          Japan        False           588          JAPAN
IN           India        False            18          INDIA
RU          Russia         True           200         RUSSIA
MOR        Morocco         True            70        MOROCCO
EG           Egypt         True            45          EGYPT


<p>Using iterrows() to iterate over every observation of a Pandas DataFrame is easy to understand, but not very efficient. On every iteration, you're creating a new Pandas Series.<br>

If you want to add a column to a DataFrame by calling a function on another column, the iterrows() method in combination with a for loop is not the preferred way to go. Instead, you'll want to use apply().<br>

Compare the iterrows() version with the apply() version to get the same result in the brics DataFrame:<br>

for lab, row in brics.iterrows() :<br>
    brics.loc[lab, "name_length"] = len(row["country"])<br>

brics["name_length"] = brics["country"].apply(len)<br>
We can do a similar thing to call the upper() method on every name in the country column. However, upper() is a method, so we'll need a slightly different approach:</p>

<li>Replace the for loop with a one-liner that uses .apply(str.upper). The call should give the same result: a column COUNTRY should be added to cars, containing an uppercase version of the country names.</li>
<li>As usual, print out cars to see the fruits of your hard labor</li>

In [25]:
# Use .apply(str.upper)
cars["COUNTRY"] = cars["country"].apply(str.upper)
print(cars)

           country drives_right  cars_per_cap        COUNTRY
US   United States         True           809  UNITED STATES
AUS      Australia        False           731      AUSTRALIA
JPN          Japan        False           588          JAPAN
IN           India        False            18          INDIA
RU          Russia         True           200         RUSSIA
MOR        Morocco         True            70        MOROCCO
EG           Egypt         True            45          EGYPT
